### DISCUSSION TITLE
Is downloading a file in bot blocking operation?

### DISCUSSION THREAD

**Discussion by Woxazu** [aiogram/aiogram#764] [Q&A]
When viewing source (applied below) code I saw that if I choose to download file to the disk bot using plain 'open' function. AFAIK this is blocking operation and none of the tasks will working while file is downloading and writing to disk. Am i right?
```
        dest = destination if isinstance(destination, io.IOBase) else open(destination, 'wb')
        async with self.session.get(url, timeout=timeout, proxy=self.proxy, proxy_auth=self.proxy_auth) as response:
            while True:
                chunk = await response.content.read(chunk_size)
                if not chunk:
                    break
                dest.write(chunk)
                dest.flush()
        if seek:
            dest.seek(0)
        return dest
```

**Comment by Olegt0rr** [comment]
`io.IOBase` is a byte-stream object
This way you can download the file to the memory and then write it to your disk without blocking (e.g. with `aiofile` lib or `loop.run_in_executor` method)
